File: selenium-project/src/test/java/com/master/seleniumproject/BooksApiTest.java
package com.master.seleniumproject;

import com.master.seleniumproject.models.Book;
import com.master.seleniumproject.models.UserContext;
import com.master.seleniumproject.steps.api.books.BooksApiSteps;
import org.testng.ITestResult;
import org.testng.annotations.AfterMethod;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

import java.util.UUID;

import static com.master.seleniumproject.config.EnvConfigs.PASSWORD;
import static com.master.seleniumproject.config.EnvConfigs.USERNAME;

public class BooksApiTest {
    private BooksApiSteps booksApiSteps;
    private UserContext userContext;
    private Book bookToCreate;
    private Book bookToUpdate;
    private Book newlyUpdatedBook;
    private Book bookToDelete;

    @BeforeMethod
    public void setUp(ITestResult result) {
        booksApiSteps = new BooksApiSteps();
        userContext = new UserContext(USERNAME, PASSWORD);
        if ("createBook".equals(result.getMethod().getMethodName())) {
            String bookTitle = "se-api-create-book-" + UUID.randomUUID().toString().substring(0, 8);
            bookToCreate = new Book(bookTitle, "George Heineman", "Study", 2021);
        }
        if ("updateBook".equals(result.getMethod().getMethodName())) {
            String bookTitle = "se-api-update-book-" + UUID.randomUUID().toString().substring(0, 8);
            bookToUpdate = new Book(bookTitle, "Markus Eisele, Natale Vinto", "Study", 2023);
            booksApiSteps.createBook(userContext, bookToUpdate);
        }
        if ("deleteBook".equals(result.getMethod().getMethodName())) {
            String bookTitle = "se-api-delete-book-" + UUID.randomUUID().toString().substring(0, 8);
            bookToDelete = new Book(bookTitle, "Boni Garcia", "Study", 2022);
            booksApiSteps.createBook(userContext, bookToDelete);
        }
    }

    @AfterMethod
    public void tearDown(ITestResult result) {
        if ("createBook".equals(result.getMethod().getMethodName())) {
            booksApiSteps.deleteBookByTitle(userContext, bookToCreate.getTitle());
        }
        if ("updateBook".equals(result.getMethod().getMethodName())) {
            booksApiSteps.deleteBookByTitle(userContext, bookToUpdate.getTitle());
            booksApiSteps.deleteBookByTitle(userContext, newlyUpdatedBook.getTitle());
        }
        if ("deleteBook".equals(result.getMethod().getMethodName())) {
            booksApiSteps.deleteBookByTitle(userContext, bookToDelete.getTitle());
        }
    }

    @Test
    public void createBook() {
        booksApiSteps.createBook(userContext, bookToCreate);
        booksApiSteps.verifyBookIsPresent(userContext, bookToCreate);
    }

    @Test
    public void updateBook() throws Exception {
        booksApiSteps.verifyBookIsPresent(userContext, bookToUpdate);
        newlyUpdatedBook = new Book(bookToUpdate.getTitle(), "UPDATED", bookToUpdate.getGenre(), bookToUpdate.getYear());
        booksApiSteps.updateBook(userContext, newlyUpdatedBook);
        booksApiSteps.verifyBookIsPresent(userContext, newlyUpdatedBook);
    }

    @Test
    public void deleteBook() {
        booksApiSteps.verifyBookIsPresent(userContext, bookToDelete);
        booksApiSteps.deleteBookByTitle(userContext, bookToDelete.getTitle());

        booksApiSteps.verifyBookIsAbsent(userContext, bookToDelete);
    }
}

File: playwright-project/tests/books2/booksApi.test.ts
import {customTest as test} from '@fixtures/customTest.fixture';
import {BooksApiSteps} from '@steps/api/books.api.steps';
import {v4 as uuid_v4} from 'uuid';
import {Book} from '@models/book';

test('Create book via API', async ({ userContext }) => {
    const bookTitle: string = `pw-api-create-book-${uuid_v4().slice(0, 8)}`;
    const book: Book = {
        title: bookTitle,
        author: 'Mark Twain',
        year: 1945,
        genre: 'Novel',
    };

    await BooksApiSteps.createBook(userContext, book);
    await BooksApiSteps.verifyBooksArePresent(userContext, [book]);

    await BooksApiSteps.deleteBookByTitle(userContext, book.title);
});

test('Update book via API', async ({ userContext }) => {
    const bookTitle: string = `pw-api-update-book-${uuid_v4().slice(0, 8)}`;
    const book: Book = {
        title: bookTitle,
        author: 'Stephen King',
        year: 1989,
        genre: 'Fantasy',
    };
    await BooksApiSteps.createBook(userContext, book);

    await BooksApiSteps.verifyBooksArePresent(userContext, [book]);
    const updatedBook: Book = {
        ...book,
        author: 'UPDATED',
    };
    await BooksApiSteps.updateBook(userContext, updatedBook);
    await BooksApiSteps.verifyBooksArePresent(userContext, [updatedBook]);

    await BooksApiSteps.deleteBookByTitle(userContext, book.title);
});

test('Delete book via API', async ({ userContext }) => {
    const bookTitle: string = `pw-api-delete-book-${uuid_v4().slice(0, 8)}`;
    const book: Book = {
        title: bookTitle,
        author: 'Nicholas Sparks',
        year: 2003,
        genre: 'Romantic drama',
    };
    await BooksApiSteps.createBook(userContext, book);

    await BooksApiSteps.verifyBooksArePresent(userContext, [book]);
    await BooksApiSteps.deleteBookByTitle(userContext, book.title);
    await BooksApiSteps.verifyBooksAreAbsent(userContext, [book]);

    await BooksApiSteps.deleteBookByTitle(userContext, book.title);
});